src/EventStore.Projections.Core/Standard/ByCorrelationId.cs

```diff
@@ -0,0 +1,132 @@
+using System;
+using System.Collections.Generic;
+using EventStore.Projections.Core.Messages;
+using EventStore.Projections.Core.Services;
+using EventStore.Projections.Core.Services.Processing;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using EventStore.Common.Utils;
+
+namespace EventStore.Projections.Core.Standard
+{
+    public class ByCorrelationId : IProjectionStateHandler, IProjectionCheckpointHandler
+    {
+        private readonly string _corrIdStreamPrefix;
+        private readonly string _corrIdCheckpointStream;
+
+        public ByCorrelationId(string source, Action<string, object[]> logger)
+        {
+            if (!string.IsNullOrWhiteSpace(source))
+                throw new InvalidOperationException("Empty source expected");
+            if (logger != null)
+            {
+//                logger("Index events by event type projection handler has been initialized");
+            }
+            // we will need to declare event types we are interested in
+            _corrIdStreamPrefix = "$bc-";
+            _corrIdCheckpointStream = "$bc";
+        }
+
+        public void ConfigureSourceProcessingStrategy(SourceDefinitionBuilder builder)
+        {
+            builder.FromAll();
+            builder.AllEvents();
+        }
+
+        public void Load(string state)
+        {
+        }
+
+        public void LoadShared(string state)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void Initialize()
+        {
+        }
+
+        public void InitializeShared()
+        {
+        }
+
+        public string GetStatePartition(CheckpointTag eventPosition, string category, ResolvedEvent data)
+        {
+            throw new NotImplementedException();
+        }
+
+        public string TransformCatalogEvent(CheckpointTag eventPosition, ResolvedEvent data)
+        {
+            throw new NotImplementedException();
+        }
+
+        public bool ProcessEvent(
+            string partition, CheckpointTag eventPosition, string category1, ResolvedEvent data,
+            out string newState, out string newSharedState, out EmittedEventEnvelope[] emittedEvents)
+        {
+            newSharedState = null;
+            emittedEvents = null;
+            newState = null;
+            if (data.EventStreamId != data.PositionStreamId)
+                return false;
+            if (!data.IsJson) 
+                return false;
+            var metadata = JObject.Parse(data.Metadata);
+            var indexedEventType = data.EventType;
+            
+            string correlationId = metadata["$correlationId"].Value<string>();
```

Comment:
Can we make the property key "$correlationId" configurable (similar to how category has a configurable option). Currently we use "correlationId" in our metadata.

Reply:
It ha always been $correlationId (ued by projection etc a well internally)

Reply:
Correct, we currently use our own by_correlation_id projection (which is almost identical to this one), it would just be nice to be able to switch over to the built in one instead. If it's too difficult to make this configurable similar to category, I understand.

```
fromAll().when({
  $any:(function(s,e) {
    if (e.metadata === null) return;
    var correlationId = e.metadata.correlationId;
    if (!correlationId) return;
    var formattedCorrelationId = correlationId.replace(/-/g, "");
    linkTo('$process.event.CorrelationId-' + formattedCorrelationId, e);
  })
})
```

Reply:
@Salgat please note that this has now been added.
the projection source now  accepts a json string to set a custom correlation id property, example: `{"correlationIdProperty":"$myCorrelationId"}`